code/modules/research/mechanic/component_exchanger.dm

```diff
@@ -0,0 +1,254 @@
+		return //Done, done and done, pull out
+
+	//Recurring option menu, what do we wish to do ?
+	var/interactoption = alert("Select desired operation", "RMCE V.12 Ready", "Output Information", "Remplace Component", "Finish Operation")
+
+	if(interactoption == "Finish Operation") //Simplest case, the user wants out
+
+		user.visible_message("<span class='notice'>[user] pulls \the [src] out of \the [M]'s maintenance hatch.</span>", \
+		"<span class='notice'>A fancy log-out screen appears and \the [src]'s systems shut down. You pull it out of \the [M] carefully.</span>")
+		return //Done
+
+	if(interactoption == "Output Information") //This also acts as a data dumping tool, if needed
+
+		var/dat
+		var/ratingpool //Used to give an estimation of the machine's quality rating
+		var/componentamount //Since the fucking circuit board is counted as a component, we can't use component_parts.len
+
+		dat += "<B>Scanning results for \the [M] :</B><BR><BR>"
+		if(M.component_parts.len)
+			for(var/obj/item/weapon/stock_parts/P in M.component_parts)
+				dat += "<B>Detected :</B> [P] of effective quality rating [P.rating].<BR>"
+				ratingpool += P.rating
+				componentamount++
+			if(ratingpool)
+				dat += "<BR>Effective quality rating of machine components : [ratingpool/componentamount].<BR>"
+		else
+			dat += "No components detected. Please ensure the scanning unit is still functional.<BR>" //Shouldn't happen
+		dat += "<BR><I>Note : You will be returned to the input screen shortly after this window is closed.</I>"
+
+		user.visible_message("<span class='notice'>[user] looks at \the [src]'s HUD as it starts a full data scan on \the [M].</span>", \
+		"<span class='notice'>You take a closer look at \the [src]'s HUD as it starts data scanning \the [M].</span>")
+		if(do_after(user, 50)) //Shouldn't be interrupted
+			user << browse(dat, "window=componentanal") //Send them the data, in a window
+			onclose(user, "componentanal")
+
+			spawn(30)
+				component_interaction(user, M)
+			return
+		else
+			user << "<span class='warning'>A blue screen suddenly flashes on \the [src]'s HUD. It appears the critical failure was caused by suddenly yanking it out of \the [M]'s maintenance hatch.</span>"
+
+	if(interactoption == "Remplace Component")
+
+		user.visible_message("<span class='notice'>[user] carefully fits \the [src] into \the [M] as it rattles and starts remplacing components.</span>", \
+		"<span class='notice'>\The [src]'s HUD flashes, a message appears stating it has started scanning and replacing \the [M]'s components.</span>")
+
+		for(var/obj/item/weapon/stock_parts/P in M.component_parts)
+			if(!Adjacent(M, user)) //Make sure the user doesn't move
+				user << "<span class='warning'>A blue screen suddenly flashes on \the [src]'s HUD. It appears the critical failure was caused by suddenly yanking it out of \the [M]'s maintenance hatch.</span>"
+				return
+			//Yes, an istype list. We don't have helpers for this, and this coder is not that sharp
```

Comment:
That was the best way I found of making component parts match, since there's no helper procs